gobcore/message_broker/notifications.py

```diff
@@ -0,0 +1,173 @@
+        del result_msg[NOTIFICATION_KEY]
+
+
+def add_notification(msg, notification):
+    """
+    Include a notification in the message
+
+    :param msg:
+    :param notification:
+    :return:
+    """
+    msg[NOTIFICATION_KEY] = {
+        NOTIFICATION_TYPE: notification.type,
+        NOTIFICATION_CONTENTS: notification.contents
+    }
+
+
+def get_notification(msg):
+    """
+    Returns an Event object for the given message
+
+    :param msg:
+    :return:
+    """
+    if msg.get(NOTIFICATION_TYPE) == EventNotification.type:
+        return EventNotification.from_msg(msg)
+
+
+class EventNotification():
+
+    type = "events"
+
+    def __init__(self, applied, last_event, header=None):
+        """
+        Initialize an Event Notification
+
+        :param applied: What kind of events have been applied and how many
+        :param last_event: The last eventid before and after the events have been applied
+        :param header:
+        """
+        self.contents = {
+            'applied': applied,
+            'last_event': last_event
+        }
+        self.header = header
+
+    @classmethod
+    def from_msg(cls, msg):
+        """
+        Construct an Event Notification object from a message
+
+        :param msg:
+        :return:
+        """
+        return cls(**msg[NOTIFICATION_CONTENTS], header=msg['header'])
+
+
+def _create_broadcast_exchange(channel, exchange):
+    """
+    Create the broadcast exchange if it does not yet exist
+
+    :param channel:
+    :param exchange:
+    :return:
+    """
+    _create_exchange(channel=channel, exchange=exchange, durable=True)
+
+
+def send_broadcast(exchange, notification_type, msg):
+    """
+    Send a broadcast message of a specified type (None for unspecied) on the specified exchange
+
+    :param exchange:
+    :param msg:
+    :return:
+    """
+    with pika.BlockingConnection(CONNECTION_PARAMS) as connection:
+        channel = connection.channel()
+
+        # Create exchange if it does not yet exist
+        _create_broadcast_exchange(channel=channel, exchange=exchange)
+
+        # Convert the message to json
+        json_msg = to_json(msg)
+
+        # Broadcast the message as a non-persistent message on the queue
+        channel.basic_publish(
+            exchange=exchange,
+            routing_key=notification_type,
+            properties=pika.BasicProperties(
+                delivery_mode=2  # Make messages persistent
+            ),
+            body=json_msg
+        )
+
+
+def listen_to_broadcasts(exchange, queue, notification_type=None):
+    """
+    Listen to broadcast messages on the specified exchange and queue of a specified type (None for all)
+
+    :param exchange:
+    :param queue:
+    :return:
+    """
+    with pika.BlockingConnection(CONNECTION_PARAMS) as connection:
+        channel = connection.channel()
+
+        # Create exchange and queue if they do not yet exist
+        _create_broadcast_exchange(channel=channel, exchange=exchange)
```

Comment:
Not really a broadcast exchange anymore. Think you can skip this function altogether now.

Reply:
Aangepast